doozerlib/cli/images_streams.py

```diff
@@ -592,18 +587,46 @@ def images_streams_prs(runtime, github_access_token, bug, interstitial, ignore_c
                     yellow_print(f'Unable to access upstream image {upstream_image} for {dgk}-- check whether buildconfigs are running successfully.')
                     if not ignore_missing_images:
                         raise
-                existing_images.add(upstream_image)  # Don't check this image again since it is a little slow to do so.
+                checked_upstream_images.add(upstream_image)  # Don't check this image again since it is a little slow to do so.
 
-            desired_parents.append(upstream_image)
+        desired_parents = []
 
-        parent_upstream_image = resolve_upstream_from(runtime, from_config)
-        if len(desired_parents) != len(builders) or not parent_upstream_image:
-            logger.warning('Unable to find all ART equivalent upstream images for this image')
-            continue
+        # There are two methods to find the desired parents for upstream Dockerfiles.
+        # 1. We can analyze the image_metadata "from:" stanza and determine the upstream
+        #    equivalents for streams and members. This is the most common flow of the code.
+        # 2. In rare circumstances, we need to circumvent this analysis and specify the
+        #    desired upstreams equivalents directly in the image metadata. This is only
+        #    required when the upstream Dockerfile uses something wholly different
+        #    than the downstream build. Use this method by specifying:
+        #      source.ci_alignment.streams_prs.from: [ list of full pullspecs ]
+        # We check for option 2 first
+        if streams_pr_config['from'] is not Missing:
+            desired_parents = streams_pr_config['from'].primitive()  # This should be list; so we're done.
+        else:
+            builders = from_config.builder or []
+            for builder in builders:
+                upstream_image = resolve_upstream_from(runtime, builder)
+                if not upstream_image:
+                    logger.warning(f'Unable to resolve upstream image for: {builder}')
+                    break
+                check_if_upstream_image_exists(upstream_image)
+                desired_parents.append(upstream_image)
+
+            parent_upstream_image = resolve_upstream_from(runtime, from_config)
+            if len(desired_parents) != len(builders) or not parent_upstream_image:
+                logger.warning('Unable to find all ART equivalent upstream images for this image')
+                continue
+
+            desired_parents.append(parent_upstream_image)
+
+        desired_parent_digest = calc_parent_digest(desired_parents)
+        logger.info(f'Found desired FROM state of: {desired_parents} with digest: {desired_parent_digest}')
 
         desired_ci_build_root_coordinate = None
-        if alignment_prs_config.ci_build_root is not Missing:
-            desired_ci_build_root_image = resolve_upstream_from(runtime, alignment_prs_config.ci_build_root)
+        if streams_pr_config.ci_build_root is not Missing:
```

Comment:
had not noticed this `ci_build_root` attr before, guess I need to update https://github.com/openshift/ocp-build-data-validator/pull/29